File: APIdoc/java/apidoc/src/main/java/io/point/apidoc/APIDoc.java
package io.point.apidoc;

import org.apache.http.NameValuePair;
import org.apache.http.client.ClientProtocolException;
import org.apache.http.client.entity.UrlEncodedFormEntity;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.client.utils.URIBuilder;
import org.apache.http.entity.mime.MultipartEntityBuilder;
import org.apache.http.entity.mime.content.FileBody;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.message.BasicNameValuePair;
import org.apache.http.util.EntityUtils;
import org.codehaus.jackson.JsonNode;
import org.codehaus.jackson.map.ObjectMapper;
import org.codehaus.jackson.node.ArrayNode;
import org.codehaus.jackson.node.JsonNodeFactory;
import org.codehaus.jackson.node.ObjectNode;

import java.io.File;
import java.io.IOException;
import java.net.URISyntaxException;
import java.util.*;

/**
 * User: jconley
 * Date: 5/5/2014
 */
public class APIDoc {

    private static CloseableHttpClient httpclient = HttpClients.createDefault();
    private static String BASE = "http://api.point.io/api/v2/";
    private static ObjectMapper mapper = new ObjectMapper();

    public static String authenticate(String email, String password, String apiKey) throws IOException {
        HttpPost httpPost = new HttpPost(BASE + "auth.json");

        List<NameValuePair> params = new ArrayList<NameValuePair>();
        params.add(new BasicNameValuePair("email", email));
        params.add(new BasicNameValuePair("password", password));
        params.add(new BasicNameValuePair("apiKey", apiKey));

        CloseableHttpResponse res = null;
        String sessionKey = null;
        try {
            httpPost.setEntity(new UrlEncodedFormEntity(params));
            res = httpclient.execute(httpPost);

            String resBody = EntityUtils.toString(res.getEntity());
            JsonNode jsonBody = mapper.readValue(resBody, JsonNode.class);
            sessionKey = jsonBody.get("RESULT").get("SESSIONKEY").getTextValue();
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if(res != null){
                res.close();
            }
        }
        return sessionKey;
    }

    public static ArrayNode listAccessRules(String sessionKey) throws IOException {
        HttpGet httpGet = new HttpGet(BASE + "accessrules/list.json");
        httpGet.addHeader("Authorization", sessionKey);

        CloseableHttpResponse res = null;
        ArrayNode response = null;
        try {
            res = httpclient.execute(httpGet);
            String resBody = EntityUtils.toString(res.getEntity());
//            System.out.println(resBody);
            response = (ArrayNode) mapper.readValue(resBody, JsonNode.class).get("RESULT").get("DATA");
        } catch (ClientProtocolException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            res.close();
        }

        return response;
    }

    public static ArrayNode listFolders(String sessionKey, String folderId) throws URISyntaxException, IOException {
        URIBuilder uri = new URIBuilder(BASE + "folders/list.json");
        uri.addParameter("folderId", folderId);

        HttpGet httpGet = new HttpGet(uri.build());
        httpGet.addHeader("Authorization", sessionKey);

        CloseableHttpResponse res = null;
        ArrayNode response = null;
        try {
            res = httpclient.execute(httpGet);
            String resBody = EntityUtils.toString(res.getEntity());
            response = (ArrayNode) mapper.readValue(resBody, JsonNode.class).get("RESULT").get("DATA");
        } catch (ClientProtocolException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            res.close();
        }

        return response;
    }

    public static String filePreview(String sessionKey, String folderId, String fileId, String fileName) throws URISyntaxException, IOException {
        URIBuilder uri = new URIBuilder(BASE + "folders/files/preview.json");
        uri.addParameter("folderid", folderId);
        uri.addParameter("fileid", fileId);
        uri.addParameter("filename", fileName);

        HttpGet httpGet = new HttpGet(uri.build());
        httpGet.addHeader("Authorization", sessionKey);

        CloseableHttpResponse res = null;
        String response = null;
        try {
            res = httpclient.execute(httpGet);
            String resBody = EntityUtils.toString(res.getEntity());
//            System.out.println(resBody);
            response = mapper.readValue(resBody, JsonNode.class).get("RESULT").asText();
        } catch (ClientProtocolException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            res.close();
        }

        return response;
    }

    public static String fileDownload(String sessionKey, String folderId, String fileId, String fileName) throws URISyntaxException, IOException {
        URIBuilder uri = new URIBuilder(BASE + "folders/files/download.json");
        uri.addParameter("folderid", folderId);
        uri.addParameter("fileid", fileId);
        uri.addParameter("filename", fileName);

        HttpGet httpGet = new HttpGet(uri.build());
        httpGet.addHeader("Authorization", sessionKey);

        CloseableHttpResponse res = null;
        String response = null;
        try {
            res = httpclient.execute(httpGet);
            String resBody = EntityUtils.toString(res.getEntity());
//            System.out.println(resBody);
            response = mapper.readValue(resBody, JsonNode.class).get("RESULT").asText();
        } catch (ClientProtocolException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            res.close();
        }

        return response;
    }

    public static String fileCreateLink(String sessionKey, String folderId, String fileId, String fileName, String remotePath, String containerId) throws IOException {
        HttpPost httpPost = new HttpPost(BASE + "links/create.json");
        httpPost.addHeader("Authorization", sessionKey);

        List<NameValuePair> params = new ArrayList<NameValuePair>();
        params.add(new BasicNameValuePair("shareid", folderId));
        params.add(new BasicNameValuePair("fileid", fileId));
        params.add(new BasicNameValuePair("filename", fileName));
        params.add(new BasicNameValuePair("remotepath", remotePath));
        params.add(new BasicNameValuePair("containerid", containerId));

        CloseableHttpResponse res = null;
        String response = null;
        try {
            httpPost.setEntity(new UrlEncodedFormEntity(params));
            res = httpclient.execute(httpPost);

            String resBody = EntityUtils.toString(res.getEntity());
//            System.out.println(resBody);
            JsonNode jsonBody = mapper.readValue(resBody, JsonNode.class);
            response = mapper.readValue(resBody, JsonNode.class).get("LINKURL").asText();
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if(res != null){
                res.close();
            }
        }
        return response;
    }

    public static String fileUpload(String sessionKey, String folderId, String fileId, String fileName, String fileContentsPath) throws IOException {
        HttpPost httpPost = new HttpPost(BASE + "folders/files/upload.json");
        httpPost.addHeader("Authorization", sessionKey);

        MultipartEntityBuilder body = MultipartEntityBuilder.create();
        body.addTextBody("folderid", folderId);
        body.addTextBody("fileid", fileId);
        body.addTextBody("filename", fileName);
        body.addPart("filecontents", new FileBody(new File(fileContentsPath)));
        httpPost.setEntity(body.build());

        CloseableHttpResponse res = null;
        String response = null;
        try {
            res = httpclient.execute(httpPost);

            System.out.println("Getting upload body");
            System.out.println(res.getStatusLine());
            String resBody = EntityUtils.toString(res.getEntity());
//            JsonNode jsonBody = mapper.readValue(resBody, JsonNode.class);
            response = resBody;
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if(res != null){
                res.close();
            }
        }
        return response;
    }

    public static JsonNode checkout(String sessionKey, String folderId, String fileName, String fileId) throws IOException, URISyntaxException {
        URIBuilder uri = new URIBuilder(BASE + "folders/files/checkout.json");
        uri.addParameter("folderid", folderId);
        uri.addParameter("fileid", fileId);
        uri.addParameter("filename", fileName);

        HttpGet httpGet = new HttpGet(uri.build());
        httpGet.addHeader("Authorization", sessionKey);

        CloseableHttpResponse res = null;
        JsonNode response = null;
        try {
            res = httpclient.execute(httpGet);
            String resBody = EntityUtils.toString(res.getEntity());
            response = mapper.readValue(resBody, JsonNode.class);
        } catch (ClientProtocolException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            res.close();
        }

        return response;
    }

    public static JsonNode checkin(String sessionKey, String folderId, String fileName, String fileId) throws IOException, URISyntaxException {
        URIBuilder uri = new URIBuilder(BASE + "folders/files/checkin.json");
        uri.addParameter("folderid", folderId);
        uri.addParameter("fileid", fileId);
        uri.addParameter("filename", fileName);

        HttpGet httpGet = new HttpGet(uri.build());
        httpGet.addHeader("Authorization", sessionKey);

        CloseableHttpResponse res = null;
        JsonNode response = null;
        try {
            res = httpclient.execute(httpGet);
            String resBody = EntityUtils.toString(res.getEntity());
            response = mapper.readValue(resBody, JsonNode.class);
        } catch (ClientProtocolException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            res.close();
        }

        return response;
    }

    public static ArrayNode listStorageTypes(String sessionKey) throws IOException {
        HttpGet httpGet = new HttpGet(BASE + "storagetypes/list.json");
        httpGet.addHeader("Authorization", sessionKey);

        CloseableHttpResponse res = null;
        ArrayNode response = null;
        try {
            res = httpclient.execute(httpGet);
            String resBody = EntityUtils.toString(res.getEntity());
//            System.out.println(resBody);
            response = (ArrayNode) mapper.readValue(resBody, JsonNode.class).get("RESULT").get("DATA");
        } catch (ClientProtocolException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            res.close();
        }

        return response;
    }

    public static ArrayNode getStorageSiteParams(String sessionKey, int siteTypeId) throws IOException {
        HttpGet httpGet = new HttpGet(BASE + "storagetypes/" + siteTypeId + "/params.json");
        httpGet.addHeader("Authorization", sessionKey);

        CloseableHttpResponse res = null;
        ArrayNode response = null;
        try {
            res = httpclient.execute(httpGet);
            String resBody = EntityUtils.toString(res.getEntity());
//            System.out.println(resBody);
            response = (ArrayNode) mapper.readValue(resBody, JsonNode.class).get("RESULT").get("DATA");
        } catch (ClientProtocolException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            res.close();
        }

        return response;
    }

    public static Map<String, Boolean> getDefaultFlags(){
        Map<String, Boolean> flags = new HashMap<String, Boolean>();
        flags.put("enabled", true);
        flags.put("loggingEnabled", false);
        flags.put("indexingEnabled", false);
        flags.put("revisionControl", false);
        flags.put("maxRevisions", false);
        flags.put("checkinCheckout", false);

        return flags;
    }

    public static JsonNode addStorageSite(String sessionKey, int siteTypeId, String name, Map<String, Boolean> flags, ObjectNode params) throws IOException {
        HttpPost httpPost = new HttpPost(BASE + "storagesites/create.json");
        httpPost.addHeader("Authorization", sessionKey);

        List<NameValuePair> formData = new ArrayList<NameValuePair>();
        formData.add(new BasicNameValuePair("siteTypeId", Integer.toString(siteTypeId)));
        formData.add(new BasicNameValuePair("name", name));
        for(Map.Entry<String, Boolean> flag : flags.entrySet()){
            formData.add(new BasicNameValuePair(flag.getKey(), flag.getValue() ? "1" : "0"));
        }

        formData.add(new BasicNameValuePair("siteArguments", params.toString()));

        CloseableHttpResponse res = null;
        JsonNode response = null;
        try {
            httpPost.setEntity(new UrlEncodedFormEntity(formData));
            res = httpclient.execute(httpPost);

            String resBody = EntityUtils.toString(res.getEntity());
            System.out.println(resBody);
            response = mapper.readValue(resBody, JsonNode.class);
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if(res != null){
                res.close();
            }
        }
        return response;
    }
}

File: APIdoc/scala/apidoc/src/main/scala/io/point/apidoc/APIDoc.scala
package io.point.apidoc

import dispatch._, Defaults._
import play.api.libs.json._
import com.ning.http.multipart.{StringPart, FilePart}

object APIDoc {

  lazy val BASE = "http://api.point.io/api/v2/"

  def authenticate(email: String, password: String, apiKey: String): Future[JsResult[String]] = {
    val authUrl = url(BASE + "auth.json")
      .POST
      .addParameter("email", email)
      .addParameter("password", password)
      .addParameter("apiKey", apiKey)

    Http(authUrl OK as.String).map( jsStr => (Json.parse(jsStr) \ "RESULT" \ "SESSIONKEY").validate[String])
  }

  def listAccessRules(sessionKey: String): Future[JsResult[JsArray]] = {
    val listUrl = url(BASE + "accessrules/list.json").addHeader("Authorization", sessionKey)
    Http(listUrl OK as.String).map ( jsStr => (Json.parse(jsStr) \ "RESULT" \ "DATA").validate[JsArray])
  }

  def listFolders(sessionKey: String, folderId: String): Future[JsResult[JsArray]] = {
    val listUrl = url(BASE + "folders/list.json").addHeader("Authorization", sessionKey).addQueryParameter("folderId", folderId)
    Http(listUrl OK as.String).map ( jsStr => (Json.parse(jsStr) \ "RESULT" \ "DATA").validate[JsArray])
  }

  def filePreview(sessionKey: String, folderId: String, fileId: String, fileName: String): Future[String] = {
    val previewUrl = url(BASE + "folders/files/preview.json")
                    .addHeader("Authorization", sessionKey)
                    .addQueryParameter("folderid", folderId)
                    .addQueryParameter("fileid", fileId)
                    .addQueryParameter("filename", fileName)
    Http(previewUrl OK as.String).map ( jsStr => jsStr)
  }

  def fileDownload(sessionKey: String, folderId: String, fileId: String, fileName: String): Future[String] = {
    val previewUrl = url(BASE + "folders/files/download.json")
      .addHeader("Authorization", sessionKey)
      .addQueryParameter("folderid", folderId)
      .addQueryParameter("fileid", fileId)
      .addQueryParameter("filename", fileName)
    Http(previewUrl OK as.String).map ( jsStr => jsStr)
  }

  def createLink(sessionKey: String, folderId: String, fileId: String, fileName: String, remotePath: String, containerId: String): Future[String] = {
    val authUrl = url(BASE + "auth.json").POST
                  .addHeader("Authorization", sessionKey)
                  .addParameter("shareid", folderId)
                  .addParameter("fileid", fileId)
                  .addParameter("filename", fileName)
                  .addParameter("remotepath", remotePath)
                  .addParameter("containerId", containerId)

    Http(authUrl OK as.String).map( jsStr => jsStr)
  }

  //RFC1867-encoded file upload
  def fileUpload(sessionKey: String, folderId: String, fileId: String, fileName: String, sourceFilePath: String): Future[String] = {
    val uploadUrl = url(BASE + "folders/files/upload.json").POST
                  .addHeader("Authorization", sessionKey)
                  .addBodyPart(new StringPart("folderid", folderId))
                  .addBodyPart(new StringPart("fileid", fileId))
                  .addBodyPart(new StringPart("filename", fileName))
                  .addBodyPart(new FilePart("filecontents", new java.io.File(sourceFilePath)))

    Http(uploadUrl OK as.String).map( jsStr => jsStr)
  }

  def fileCheckout(sessionKey: String, folderId: String, fileId: String, fileName: String): Future[String] = {
    val previewUrl = url(BASE + "folders/files/checkout.json")
      .addHeader("Authorization", sessionKey)
      .addQueryParameter("folderid", folderId)
      .addQueryParameter("fileid", fileId)
      .addQueryParameter("filename", fileName)
    Http(previewUrl OK as.String).map ( jsStr => jsStr)
  }

  def fileCheckin(sessionKey: String, folderId: String, fileId: String, fileName: String): Future[String] = {
    val previewUrl = url(BASE + "folders/files/checkin.json")
      .addHeader("Authorization", sessionKey)
      .addQueryParameter("folderid", folderId)
      .addQueryParameter("fileid", fileId)
      .addQueryParameter("filename", fileName)
    Http(previewUrl OK as.String).map ( jsStr => jsStr)
  }

  def listStorageTypes(sessionKey: String): Future[JsResult[JsArray]] = {
    val listUrl = url(BASE + "storagetypes/list.json").addHeader("Authorization", sessionKey)
    Http(listUrl OK as.String).map ( jsStr => (Json.parse(jsStr) \ "RESULT" \ "DATA").validate[JsArray])
  }

  def listStorageTypeParams(sessionKey: String, siteTypeId: Int): Future[JsResult[JsArray]] = {
    val listUrl = url(BASE + "storagetypes/" + siteTypeId + "/params.json").addHeader("Authorization", sessionKey)
    Http(listUrl OK as.String).map ( jsStr => (Json.parse(jsStr) \ "RESULT" \ "DATA").validate[JsArray])
  }

  def defaultFlags = Json.obj(
    "enabled" -> "1",
    "loggingEnabled" -> "0",
    "indexingEnabled" -> "0",
    "revisionControl" -> "0",
    "maxRevisions" -> "0",
    "checkinCheckout" -> "0"
  )

  def createStorageSite(sessionKey: String, siteTypeId: Int, name: String, flags: JsObject = defaultFlags, siteArguments: JsObject): Future[String] = {
    var createSiteReq = url(BASE + "auth.json").POST
      .addHeader("Authorization", sessionKey)
      .addParameter("siteTypeId", siteTypeId.toString)
      .addParameter("name", name)

    flags.fields.foreach { v =>
      createSiteReq = createSiteReq.addParameter(v._1, v._2.as[String])
    }

    siteArguments.fields.foreach { v =>
      createSiteReq = createSiteReq.addParameter(v._1, v._2.as[String])
    }

    Http(createSiteReq OK as.String).map( jsStr => jsStr)
  }
}
